File: votable.py
"""

Setup a venv and install the substrate-interface:

```bash
python3 -m venv venv
venv/bin/pip install substrate-interface
```

Now add a cron job to run every hour to check for votable referenda. You need to:
- Replace the path to the substrate-scripts folder with an absolute path.
- Replace the account address with your own.
- Add tracks that you can vote on. The tracks below are for Rank 3 fellows.

0 * * * * cd REPLACE_ME/substrate-scripts && venv/bin/python3 votable.py 16a357f5Sxab3V2ne4emGQvqJaCLeYpTMx3TCjnQhmJQ71DX --tracks 1 2 3 11 21
"""

import os
import json
import math
import time
import argparse

from substrateinterface import SubstrateInterface, Keypair
from substrateinterface.exceptions import SubstrateRequestException

def main():
	parser = argparse.ArgumentParser(description='List votable referenda')
	parser.add_argument('account', help='The account to check')
	parser.add_argument('--tracks', help='The tracks to check', nargs='+', default=[])
	args = parser.parse_args()
	tracks = [int(t) for t in args.tracks]
	ACC = args.account

	if len(tracks) == 0:
		print("Please specify the tracks to check with --tracks")
		exit(1)

	chain = SubstrateInterface(
		url="wss://polkadot-collectives-rpc.polkadot.io",
		#url="ws://127.0.0.1:8000",
	)
	print(f"Connected to {chain.name}: {chain.chain} v{chain.version}")

	good_state = ['DecisionDepositPlaced', 'Ongoing', 'Submitted', 'DecisionStarted', 'ConfirmStarted']

	ongoing = []

	# List all referenda
	referenda = chain.query_map("FellowshipReferenda", "ReferendumInfoFor")
	referenda_by_index = {}
	for index, ref in referenda:
		ref = ref.value
		index = index.value
		
		status = list(ref.keys())[0]
		if not status in good_state:
			continue
		
		ongoing.append(index)
		referenda_by_index[index] = ref[status]

	# Now get all referenda from the collective pallet:
	voted = chain.query_map("FellowshipCollective", "Voting")
	for (index, account), vote in voted:
		if not index in ongoing:
			continue
		
		if account == ACC:
			ongoing.remove(index)

	can_vote = []
	for i in ongoing:
		origin = referenda_by_index[i]['origin']['FellowshipOrigins']
		track = referenda_by_index[i]['track']

		if track in tracks:
			print(f"[ {origin} ] Please vote on https://collectives.subsquare.io/fellowship/referenda/{i}")
			can_vote.append(i)

	if len(can_vote) == 0:
		print("No referenda to vote on")
	else:
		found = len(can_vote)
		print(f"Found {found} referenda to vote on")
		as_str = ", ".join([str(i) for i in can_vote])
		os.system(f"osascript -e 'display notification \"Please vote on referenda: {as_str}\" with title \"New Fellowship Referenda\" sound name \"Submarine\"'")

if __name__ == "__main__":
	main()

File: papi/votable.js
import { ApiPromise, WsProvider } from '@polkadot/api';
import { Command } from 'commander';
import { exec } from "child_process";
import { startVotingServer } from './votable-ui.js';
import https from 'https';
import dotenv from 'dotenv';

dotenv.config({ path: '../.env' });

async function didVote(refIndex, account) {
  const options = {
    method: 'POST',
    hostname: 'polkadot.api.subscan.io',
    path: '/api/scan/referenda/votes',
    headers: {
      'User-Agent': 'Apidog/1.0.0 (https://apidog.com)',
      'Content-Type': 'application/json',
      'X-API-Key': process.env.API_KEY
    }
  };

  const body = JSON.stringify({
    account: account,
    order: "asc",
    page: 0,
    referendum_index: refIndex,
    row: 10,
    sort: "conviction",
    valid: "valid"
  });

  return new Promise((resolve, reject) => {
    const req = https.request(options, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        try {
          const votes = JSON.parse(data).data;
          if (votes.count == 0) {
            resolve(false);
          } else {
            resolve(true);
          }
        } catch (error) {
          reject(new Error('Error parsing JSON response'));
        }
      });
    });

    req.on('error', (error) => {
      reject(error);
    });

    req.write(body);
    req.end();
  });
}

async function main() {
  const program = new Command();
  program
    .argument('<account>', 'The account to check')
    .option('--tracks <tracks...>', 'The tracks to check', [])
    .option('--all-tracks', 'Check all tracks')
    .option('--rpc <string>', 'The RPC endpoint to connect to', 'wss://polkadot-collectives-rpc.polkadot.io')
    .option('--pallet <string>', 'The pallet to check', 'fellowshipReferenda')
    .parse(process.argv);

  const options = program.opts();
  const ACC = program.args[0];
  const tracks = options.tracks.map(Number);
  const allTracks = options.allTracks;

  if (tracks.length === 0 && !allTracks) {
    console.log("Please specify the tracks to check with --tracks or use --all-tracks");
    process.exit(1);
  }

  console.log(`Connecting to ${options.rpc}`);
  const wsProvider = new WsProvider(options.rpc);
  const api = await ApiPromise.create({ provider: wsProvider });

  console.log(`Connected to ${api.runtimeChain}: ${api.runtimeVersion.specName} v${api.runtimeVersion.specVersion}`);

  const goodStates = ['decisionDepositPlaced', 'ongoing', 'submitted', 'decisionStarted', 'confirmStarted'];

  let ongoing = [];
  const referendaByIndex = {};

  // List all referenda
  const referenda = await api.query[options.pallet].referendumInfoFor.entries();
  for (const [key, ref] of referenda) {
    const index = key.args[0].toNumber();
    const status = ref.value.toJSON();
    const statusKey = Object.keys(status)[0];

    if (!goodStates.includes(statusKey))
		continue;

    ongoing.push(index);
    referendaByIndex[index] = status[statusKey];
  }

  // Now get votes and filter out the ones that have already voted
  var canVote = [];
  for (const index of ongoing) {
    const voted = await didVote(index, ACC);

    if (!voted) {
      canVote.push(index);
      console.log(`You can vote on referendum ${index}`);
      if (canVote.length > 3)
        break;
    }
  }

  if (canVote.length === 0) {
    console.log("No referenda to vote on");
    process.exit(0);
  } else {
    // 30 minutes to vote
    setTimeout(function () {
      process.exit(0);
    }, 3 * 600000);
    const found = canVote.length;
    console.log(`Found ${found} referenda to vote on`);
    const asStr = canVote.join(", ");
    const plural = found === 1 ? "um" : "a";
    console.log(`Please vote on referend${plural}: ${asStr}`);
    exec(`osascript -e 'display notification "Please vote on ${found} OpenGov referend${plural}" with title "Votable referend${plural}"'`);
    
    startVotingServer(canVote);
  }

  await api.disconnect();
}

main()
